proposals/p2360.md

```diff
@@ -0,0 +1,478 @@
+    -   Similarly, `{}` is of type `{} as type`, which is of type `type`.
+    -   More generally, _the type of any type is `type`_.
+-   Given `T:! Hashable`, `T` is formally not a type, because it is not of type
+    `type`. Indeed, member lookup into `type`, and hence into types, finds
+    nothing, but member lookup into `T` finds members of `Hashable`.
+-   Contexts in which a type is required, such as on the right hand side of a
+    `:` or `:!` binding or the return type of a function, perform an implicit
+    conversion to type `type`.
+    -   This means that a user-defined type can implement `ImplicitAs(type)`,
+        annotated in whatever way we eventually decide to annotate functions
+        that are usable at compile time, and values of that type will be usable
+        as types, in the same way that `()` and `{}` are.
+-   There are no singleton type-of-types. It's still possible to have, for
+    example, the type of `i32` be a type whose only value is `i32`. The value
+    `i32` would then not formally be a type, as its type would not be `type`,
+    but it could still implicitly convert to `type`, as `()` does, if we so
+    desire.
+    -   This extends to other cases: we could say that for integer literals,
+        which we have recently been describing as having type `IntLiteral(n)`,
+        there is an `impl IntLiteral(n) as ImplicitAs(type)` which converts a
+        value of type `IntLiteral(n)` to the `type` value `IntLiteral(n)`. This
+        would then permit `let v: 0 = 0;`.
+    -   Such changes to `i32` or integer literals are not part of this proposal.
+
+## Details
+
+### Terminology
+
+The changes to the language rules in this proposal are fairly small, but there
+is a significant change in how we describe those rules. We now use the following
+terminology:
+
+-   A _type_ is a value of type `type`.
+-   A _facet type_ is a constrained or unconstrained version of `type`, such as
+    an interface type or a named constraint type. We previously referred to a
+    facet type as a type-of-type, but that is no longer accurate, as the type of
+    any type is now by definition `type`.
+-   A _facet_ is a value of a facet type. For example, `i32 as Hashable` is a
+    facet, and `Hashable` is a facet type. Note that all types are facets.
+    -   A facet value can be thought of as a tuple of a possibly-symbolic
+        identity for a type, plus any witnesses needed to demonstrate that the
+        type satisfies the constraints of the facet type. A type is then a facet
+        value that has no witnesses.
+-   We use the term _generic type_ to refer to a type or facet introduced by a
+    `:!` binding, such as in a generic parameter or associated constant.
+    -   This gives rise to terms such as _generic type parameter_ and
+        _associated generic type_.
+    -   It is worth noting that a generic type is not necessarily a type, as its
+        type may be an interface or constraint rather than `type`.
```

Comment:
Maybe slightly clearer to use the terminology above?

```suggestion
    -   It is worth noting that a generic type is not necessarily a type. It may
        be a facet, and its type may be a facet type rather than `type`.
```

Reply:
```suggestion
     -   It is worth noting that a generic type is not necessarily a type. It may
        be a facet, and its type may be a facet type other than `type`.
```
(rather -> other, because `type` is a facet type.)